Extra api call retries for 422

**Describe the bug**
When we fail to create an environment and we get 422 we retry it although there is no need as this is not server side issue


**To Reproduce**
```
terraform {
  required_providers {
    env0 = {
      source = "env0/env0"
    }
  }
}

provider "env0" {
}

resource "random_uuid" "workspace" {
}

resource "env0_environment" "example" {
  name                = "environment"
  project_id          = "."
  template_id         = "."
  prevent_auto_deploy = true

  sub_environment_configuration {
    alias = "first-alias"
    configuration {
      name          = "broken-hcl"
      value         = "{"
      schema_format = "HCL"
    }
    revision = "master"
    workspace = random_uuid.workspace.result
  }
}
```
Logs - 
```
2024-09-05T15:08:08.841+0200 [INFO]  provider.terraform-provider-env0_v1.20.7: Received response: @module=provider.env0_api_client method=POST status="422 Unprocessable Entity" url=https://api.env0.com/environments @caller=github.com/env0/terraform-provider-env0/env0/provider.go:199 timestamp="2024-09-05T15:08:08.840+0200"
2024-09-05T15:08:09.842+0200 [INFO]  provider.terraform-provider-env0_v1.20.7: Sending request: method=POST url=/environments @caller=github.com/env0/terraform-provider-env0/env0/provider.go:194 @module=provider.env0_api_client timestamp="2024-09-05T15:08:09.841+0200"
2024-09-05T15:08:12.948+0200 [INFO]  provider.terraform-provider-env0_v1.20.7: Received response: url=https://api.env0.com/environments @caller=github.com/env0/terraform-provider-env0/env0/provider.go:199 @module=provider.env0_api_client method=POST status="422 Unprocessable Entity" timestamp="2024-09-05T15:08:12.947+0200"
2024-09-05T15:08:14.102+0200 [INFO]  provider.terraform-provider-env0_v1.20.7: Sending request: @module=provider.env0_api_client method=POST url=/environments @caller=github.com/env0/terraform-provider-env0/env0/provider.go:194 timestamp="2024-09-05T15:08:14.102+0200"
env0_environment.example: Still creating... [10s elapsed]
2024-09-05T15:08:17.151+0200 [INFO]  provider.terraform-provider-env0_v1.20.7: Received response: url=https://api.env0.com/environments @caller=github.com/env0/terraform-provider-env0/env0/provider.go:199 @module=provider.env0_api_client method=POST status="422 Unprocessable Entity" timestamp="2024-09-05T15:08:17.150+0200"
2024-09-05T15:08:19.515+0200 [INFO]  provider.terraform-provider-env0_v1.20.7: Sending request: @caller=github.com/env0/terraform-provider-env0/env0/provider.go:194 @module=provider.env0_api_client method=POST url=/environments timestamp="2024-09-05T15:08:19.515+0200"
2024-09-05T15:08:22.416+0200 [INFO]  provider.terraform-provider-env0_v1.20.7: Received response: @module=provider.env0_api_client method=POST status="422 Unprocessable Entity" url=https://api.env0.com/environments @caller=github.com/env0/terraform-provider-env0/env0/provider.go:199 timestamp="2024-09-05T15:08:22.416+0200"
env0_environment.example: Still creating... [20s elapsed]
2024-09-05T15:08:26.135+0200 [INFO]  provider.terraform-provider-env0_v1.20.7: Sending request: url=/environments @caller=github.com/env0/terraform-provider-env0/env0/provider.go:194 @module=provider.env0_api_client method=POST timestamp="2024-09-05T15:08:26.134+0200"
2024-09-05T15:08:29.092+0200 [INFO]  provider.terraform-provider-env0_v1.20.7: Received response: status="422 Unprocessable Entity" url=https://api.env0.com/environments @caller=github.com/env0/terraform-provider-env0/env0/provider.go:199 @module=provider.env0_api_client method=POST timestamp="2024-09-05T15:08:29.092+0200"
2024-09-05T15:08:33.686+0200 [INFO]  provider.terraform-provider-env0_v1.20.7: Sending request: @caller=github.com/env0/terraform-provider-env0/env0/provider.go:194 @module=provider.env0_api_client method=POST url=/environments timestamp="2024-09-05T15:08:33.686+0200"
env0_environment.example: Still creating... [30s elapsed]
2024-09-05T15:08:36.584+0200 [INFO]  provider.terraform-provider-env0_v1.20.7: Received response: status="422 Unprocessable Entity" url=https://api.env0.com/environments @caller=github.com/env0/terraform-provider-env0/env0/provider.go:199 @module=provider.env0_api_client method=POST timestamp="2024-09-05T15:08:36.584+0200"
2024-09-05T15:08:36.585+0200 [ERROR] provider.terraform-provider-env0_v1.20.7: Response contains error diagnostic: @caller=github.com/hashicorp/terraform-plugin-go@v0.20.0/tfprotov5/internal/diag/diagnostics.go:62 @module=sdk.proto tf_resource_type=env0_environment tf_proto_version=5.4 tf_req_id=11c9bfa9-c3ff-5cfb-952e-336a195aebbf tf_rpc=ApplyResourceChange diagnostic_detail="" diagnostic_severity=ERROR diagnostic_summary="could not create environment: 422 Unprocessable Entity: Error in variable arel: failed to parse content as HCL: parse config: [HCL:1,6-7: Unclosed configuration block; There is no closing brace for this block before the end of the file. This may be caused by incorrect brace nesting elsewhere in this file.]" tf_provider_addr=provider timestamp="2024-09-05T15:08:36.585+0200"
```

**Expected behavior**
It should only send one API call

**Provider Version**
1.20.7

**Additional context**
Add any other context about the problem here.

--- a/env0/provider.go
+++ b/env0/provider.go
@@ -199,7 +199,6 @@ func configureProvider(version string, p *schema.Provider) schema.ConfigureConte
 				tflog.SubsystemInfo(subCtx, "env0_api_client", "Received response", map[string]interface{}{"method": r.Request.Method, "url": r.Request.URL, "status": r.Status()})
 				return nil
 			}).
-			AddRetryAfterErrorCondition().
 			AddRetryCondition(func(r *resty.Response, err error) bool {
 				if r == nil {
 					// No response. Possibly a networking issue (E.g. DNS lookup failure).

--- a/env0/provider.go
+++ b/env0/provider.go
@@ -171,56 +171,55 @@ func Provider(version string) plugin.ProviderFunc {
 	}
 }
 
+func createRestyClient(ctx context.Context) *resty.Client {
+	isIntegrationTest := false
+
+	if os.Getenv("INTEGRATION_TESTS") == "1" {
+		isIntegrationTest = true
+	}
+
+	subCtx := tflog.NewSubsystem(ctx, "env0_api_client")
+
+	return resty.New().SetRetryCount(5).
+		SetRetryWaitTime(time.Second).
+		SetRetryMaxWaitTime(time.Second * 5).
+		OnBeforeRequest(func(c *resty.Client, r *resty.Request) error {
+			if r != nil {
+				tflog.SubsystemInfo(subCtx, "env0_api_client", "Sending request", map[string]interface{}{"method": r.Method, "url": r.URL})
+			}
+			return nil
+		}).
+		OnAfterResponse(func(c *resty.Client, r *resty.Response) error {
+			tflog.SubsystemInfo(subCtx, "env0_api_client", "Received response", map[string]interface{}{"method": r.Request.Method, "url": r.Request.URL, "status": r.Status()})
+			return nil
+		}).
+		AddRetryCondition(func(r *resty.Response, err error) bool {
+			if r == nil {
+				// No response. Possibly a networking issue (E.g. DNS lookup failure).
+				tflog.SubsystemWarn(subCtx, "env0_api_client", "No response, retrying request")
+				return true
+			}
+
+			// When running integration tests 404 may occur due to "database eventual consistency".
+			// Retry when there's a 5xx error. Otherwise do not retry.
+			if r.StatusCode() >= 500 || (isIntegrationTest && r.StatusCode() == 404) {
+				tflog.SubsystemWarn(subCtx, "env0_api_client", "Received a failed or not found response, retrying request", map[string]interface{}{"method": r.Request.Method, "url": r.Request.URL, "status code": r.StatusCode()})
+				return true
+			}
+
+			if r.StatusCode() == 200 && isIntegrationTest && r.String() == "[]" {
+				tflog.SubsystemWarn(subCtx, "env0_api_client", "Received an empty list , retrying request", map[string]interface{}{"method": r.Request.Method, "url": r.Request.URL})
+				return true
+			}
+
+			return false
+		})
+}
+
 func configureProvider(version string, p *schema.Provider) schema.ConfigureContextFunc {
 	userAgent := p.UserAgent("terraform-provider-env0", version)
 
 	return func(ctx context.Context, d *schema.ResourceData) (interface{}, diag.Diagnostics) {
-		restyClient := resty.New()
-
-		isIntegrationTest := false
-
-		if os.Getenv("INTEGRATION_TESTS") == "1" {
-			isIntegrationTest = true
-		}
-
-		subCtx := tflog.NewSubsystem(ctx, "env0_api_client")
-
-		restyClient = restyClient.
-			SetRetryCount(5).
-			SetRetryWaitTime(time.Second).
-			SetRetryMaxWaitTime(time.Second * 5).
-			OnBeforeRequest(func(c *resty.Client, r *resty.Request) error {
-				if r != nil {
-					tflog.SubsystemInfo(subCtx, "env0_api_client", "Sending request", map[string]interface{}{"method": r.Method, "url": r.URL})
-				}
-				return nil
-			}).
-			OnAfterResponse(func(c *resty.Client, r *resty.Response) error {
-				tflog.SubsystemInfo(subCtx, "env0_api_client", "Received response", map[string]interface{}{"method": r.Request.Method, "url": r.Request.URL, "status": r.Status()})
-				return nil
-			}).
-			AddRetryCondition(func(r *resty.Response, err error) bool {
-				if r == nil {
-					// No response. Possibly a networking issue (E.g. DNS lookup failure).
-					tflog.SubsystemWarn(subCtx, "env0_api_client", "No response, retrying request")
-					return true
-				}
-
-				// When running integration tests 404 may occur due to "database eventual consistency".
-				// Retry when there's a 5xx error. Otherwise do not retry.
-				if r.StatusCode() >= 500 || (isIntegrationTest && r.StatusCode() == 404) {
-					tflog.SubsystemWarn(subCtx, "env0_api_client", "Received a failed or not found response, retrying request", map[string]interface{}{"method": r.Request.Method, "url": r.Request.URL, "status code": r.StatusCode()})
-					return true
-				}
-
-				if r.StatusCode() == 200 && isIntegrationTest && r.String() == "[]" {
-					tflog.SubsystemWarn(subCtx, "env0_api_client", "Received an empty list , retrying request", map[string]interface{}{"method": r.Request.Method, "url": r.Request.URL})
-					return true
-				}
-
-				return false
-			})
-
 		apiKey, ok := d.GetOk("api_key")
 		if !ok {
 			return nil, diag.Diagnostics{diag.Diagnostic{Severity: diag.Error, Detail: `The argument "api_key" is required, but no definition was found.`}}
@@ -236,7 +235,7 @@ func configureProvider(version string, p *schema.Provider) schema.ConfigureConte
 			ApiSecret:   apiSecret.(string),
 			ApiEndpoint: d.Get("api_endpoint").(string),
 			UserAgent:   userAgent,
-			RestClient:  restyClient,
+			RestClient:  createRestyClient(ctx),
 		})
 		if err != nil {
 			return nil, diag.Diagnostics{diag.Diagnostic{Severity: diag.Error, Summary: err.Error()}}
@@ -248,6 +247,7 @@ func configureProvider(version string, p *schema.Provider) schema.ConfigureConte
 		if _, err := apiClient.OrganizationId(); err != nil {
 			return nil, diag.Diagnostics{diag.Diagnostic{Severity: diag.Error, Summary: err.Error()}}
 		}
+
 		return apiClient, nil
 	}
 }

--- a/env0/provider_test.go
+++ b/env0/provider_test.go
@@ -3,18 +3,29 @@ package env0
 import (
 	"context"
 	"fmt"
+	"net/http"
 	"os"
 	"strings"
 	"testing"
 
 	"github.com/env0/terraform-provider-env0/client"
 	"github.com/env0/terraform-provider-env0/utils"
+	"github.com/go-resty/resty/v2"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+	"github.com/jarcoal/httpmock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
 	"go.uber.org/mock/gomock"
 )
+
+type testRestyClientSuite struct {
+	suite.Suite
+	client *resty.Client
+	url    string
+}
 
 func runUnitTest(t *testing.T, testCase resource.TestCase, mockFunc func(mockFunc *client.MockApiClientInterface)) {
 	t.Helper()
@@ -112,4 +123,71 @@ func TestMissingConfigurations(t *testing.T) {
 	for expectedError, envVars := range envVarsTestCases {
 		testMissingEnvVar(t, envVars, expectedError)
 	}
+}
+
+func (suite *testRestyClientSuite) SetupTest() {
+	httpmock.Reset()
+}
+
+func (suite *testRestyClientSuite) SetupSuite() {
+	httpmock.ActivateNonDefault(suite.client.GetClient())
+}
+
+func (suite *testRestyClientSuite) TearDownAllSuite() {
+	httpmock.Deactivate()
+}
+
+func (suite *testRestyClientSuite) TestOkResponse() {
+	t := suite.T()
+
+	httpmock.RegisterResponder("GET", suite.url, httpmock.NewStringResponder(http.StatusOK, "OK"))
+
+	res, err := suite.client.R().Get(suite.url)
+
+	if assert.NoError(t, err) {
+		assert.Equal(t, http.StatusOK, res.StatusCode())
+		assert.Equal(t, "OK", res.String())
+	}
+
+	assert.Equal(t, 1, httpmock.GetTotalCallCount())
+}
+
+func (suite *testRestyClientSuite) Test4xxResponse() {
+	t := suite.T()
+
+	httpmock.RegisterResponder("GET", suite.url, httpmock.NewStringResponder(http.StatusBadRequest, "BAD"))
+
+	res, err := suite.client.R().Get(suite.url)
+
+	if assert.NoError(t, err) {
+		assert.Equal(t, http.StatusBadRequest, res.StatusCode())
+		assert.Equal(t, "BAD", res.String())
+	}
+
+	// Should be called once - no retries.
+	assert.Equal(t, 1, httpmock.GetTotalCallCount())
+}
+
+func (suite *testRestyClientSuite) Test5xxResponse() {
+	t := suite.T()
+
+	httpmock.RegisterResponder("GET", suite.url, httpmock.NewStringResponder(http.StatusInternalServerError, "BAD"))
+
+	res, err := suite.client.R().Get(suite.url)
+
+	if assert.NoError(t, err) {
+		assert.Equal(t, http.StatusInternalServerError, res.StatusCode())
+		assert.Equal(t, "BAD", res.String())
+	}
+
+	// Should be called multiple times - retries.
+	assert.Equal(t, 6, httpmock.GetTotalCallCount())
+}
+
+func TestRestyClientSuite(t *testing.T) {
+	s := &testRestyClientSuite{
+		client: createRestyClient(context.Background()),
+		url:    "http://fake.env0.com/fake",
+	}
+	suite.Run(t, s)
 }

